docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,410 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #29.        r = m * (1 - ((min_dist ** 2)  / (2 * m))) 
</pre><p>In STUMPY, we need to support non-normalized Euclidean distances (i.e., without z-normalization). How would we need to transform <code>min_dist</code> in that case (i.e., what is <code>r</code>  when we are dealing with non-normalized subsequences)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Reply:
Please note that the parameter `r` used in the original paper is different than the `r` used here. In fact, the `r` used in the original paper is the `min_dist` we used here. The original paper used the `min_dist` and they only consider z-normalized case. 

While the author of MATLAB code of `MELRIN` (i.e. original version) used `min_dist`, the author of MATLAB code of `fastMERLIN`  used ` r = m * (1 - ((min_dist ** 2)  / (2 * m))) ` to get advantage from the dot product of subsequences. However, they didn't discuss the performance for non-normalized sequences. 

**Idea:**
we can use `min_dist` directly (same as what the original paper did) and see how it performs for non-normalized case. I also think the initial value for `min_dist` in non-normalized case can be calculated as: `(np.max(T) - np.min(T)) * np.sqrt(L)`, where T is the main time series. So, there is no need to be worried about that (for now).

---
**NOTE:**
Since you said we should try to be close to the implementation of the original paper, and you also said the STUMPY should work for both normalized and non-normalized cases, I can create another function for non-normalized sequences. In this function, we can use `min_dist` rather than `r`. We cannot use `np.matmul()` anymore; however, we are following exactly what provided in the Table 1 of paper, i.e. creating `nested for-loop`. The inner one iterates through the so-far-selected candidates.

Reply:
<p>Do you know how they derived the relationship for <code>r</code>?   Why is it <code>r = m * (1 - ((min_dist ** 2)  / (2 * m)))</code> ?</p>

Reply:
As provided in [Pearson notebook](https://github.com/TDAmeritrade/stumpy/blob/main/docs/Pearson.ipynb), there is a relationship between `ED(z-norm)` and `Pearson Correlation (PC)` (see formula (1)). And, as explain in this [paper](https://arxiv.org/pdf/1601.02213.pdf), the `PC` of two z-normalized time series is `(1 / len(T)) * inner product` (see page 3).

So, if we consider the z-normlized Euclidean distance `ED(z-norm) = min_dist`, the value of `r` (which corresponds to the inner product) becomes:
`r = m * (1 - ((min_dist ** 2) / (2 * m)))`

Reply:
Okay, can you follow the Pearson notebook (which uses LaTex in Markdown) to show the full derivation in this notebook?

Also, I don't know if this is helpful but there is certainly a dot-product (matmul) relationship with non-normalized Euclidean distance:

```
(x - y) ** 2 = x**2 + y**2 - 2*x*y
```

Though, I don't know what `min_dist` is for this

Reply:
> Okay, can you follow the Pearson notebook (which uses LaTex in Markdown) to show the full derivation in this notebook?

Will do so. I am still waiting to go through  all comments and get into agreement first (to avoid the issues we faced before), and then I will apply changes in the notebook and push the notebook to the repo.


> Also, I don't know if this is helpful but there is certainly a dot-product (matmul) relationship with non-normalized Euclidean distance:

I can go ahead and "try" to use this for non-normalized sequence. 
**Just a thought:** Let's say we have a candidate and we already calculate its distance with subsequence `T[i:i+m]` at `i-th iteration`. IF that candidate remains for the next iteration (i.e. not removed), we can calculate its distance to `T[i+1:i+1+m]`  (in the next iteration) by using the result of previous iteration. Maybe we can use this concept for non-normalized sequences in `MERLIN`.


> Though, I don't know what `min_dist` is for this

Well, I think it is: `sqrt( m * ((max(T) - min(T)^2) )`. (Because, the maximum difference in each time stamp is `max(T) - min(T)`

Reply:
> Will do so. I am still waiting to go through all comments and get into agreement first (to avoid the issues we faced before), and then I will apply changes in the notebook and push the notebook to the repo.

Sounds good

> IF that candidate remains for the next iteration (i.e. not removed), we can calculate its distance to T[i+1:i+1+m] (in the next iteration) by using the result of previous iteration.

This is similar to the concept behind STOMP. Let's think about it after we wrap things up

> Well, I think it is: sqrt( m * ((max(T) - min(T)^2) ).

Please prove it to me with a derivation. It would be useful to document this as I have for the Pearson notebook,

Reply:
I will provide the proof for the `r` (in normalized case) and `min_dist` (in non-normalized case) in the notebook.  You can check them out after the next push of commits.